### PR TITLE
"a large number of fixes and cleanups"

### PR DESCRIPTION
**Pull request by mnunberg**

### CODE REVIEW

**Comment by mnunberg** [comment]
Can this be merged? (this is the same as what you saw last week. All I did was add 'NOSTEM' support).

**Comment by dvirsky** [comment]
@mnunberg the only thing I'm not sure about is the error map. 

How about this: we create our own error object, let's call it IndexingError or something, and we return that. If no error occurred - we return nil. If at least one document has erred - we return our error ojbect,  and it contains a slice of sub errors (not a map since the access is by sequence anyway), and you can inspect that. i.e.

```go
type BulkIndexingError struct {
    errors []error
}

func (e *BulkIndexingError)Error() string {
    // yada yada
}

func (e *BulkIndexingError)Errors() []error {
   return e.errors
}

```

And thus we can keep the multi indexing function with the old signature which is more idiomatic 


```go
func (i *Client) IndexOptions(opts IndexingOptions, docs ...Document) error {
}
```

File: redisearch/client.go [hunk 1]
```diff
@@ -307,3 +314,153 @@ func (i *Client) Drop() error {
...
+	TermCount            uint64  `redis:"num_terms"`
+	MaxDocID             uint64  `redis:"max_doc_id"`
+	InvertedIndexSizeMB  float64 `redis:"inverted_sz_mb"`
+	OffsetVectorSizeMB   float64 `redis:"offset_vector_sz_mb"`
+	DocTableSizeMB       float64 `redis:"doc_table_size_mb"`
+	KeyTableSizeMB       float64 `redis:"key_table_size_mb"`
+	RecordsPerDocAvg     float64 `redis:"records_per_doc_avg"`
+	BytesPerRecordAvg    float64 `redis:"bytes_per_record_avg"`
+	OffsetsPerTermAvg    float64 `redis:"offsets_per_term_avg"`
+	OffsetBitsPerTermAvg float64 `redis:"offset_bits_per_record_avg"`
```
Comment (dvirsky) [inline redisearch/client.go]: This will soon need the gc stats as well

**Comment by mnunberg** [comment]
Good idea about the error map, I'll get around to it soon.

File: redisearch/client.go [hunk 2]
```diff
@@ -148,6 +148,16 @@ type IndexingOptions struct {
 	Replace  bool
 }
 
+// MultiError Represents one or more errors
+type MultiError map[int]error
+
+func (e MultiError) Error() string {
```
Comment (dvirsky) [inline redisearch/client.go]: not a blocker, but wouldn't you in this case rather chain all error strings somehow?

File: redisearch/client.go [hunk 3]
```diff
@@ -148,6 +148,16 @@ type IndexingOptions struct {
 	Replace  bool
 }
 
+// MultiError Represents one or more errors
+type MultiError map[int]error
```
Comment (dvirsky) [inline redisearch/client.go]: I'd rather have a slice here where each sub error correlates to the index of the operation that triggered it

File: redisearch/client.go (hunk 2 above)
Comment (mnunberg) [inline redisearch/client.go]: Wouldn't that require line breaks to make it readable, and wouldn't line breaks confuse a lot of things?

Comment (dvirsky) [inline redisearch/client.go]: I put newlines, we can always make it look like an array. It's just for debug printing etc.

**Pull request merged**